[Cocoa] Plumb a video receiver endpoint from LinearMediaPlayer to MediaPlayerPrivate

aestes opened the pull request:
#### fcbb7a1f693fb471274776914171257eb48e08de
<pre>
[Cocoa] Plumb a video receiver endpoint from LinearMediaPlayer to MediaPlayerPrivate
<a href="https://bugs.webkit.org/show_bug.cgi?id=269490">https://bugs.webkit.org/show_bug.cgi?id=269490</a>
<a href="https://rdar.apple.com/123028396">rdar://123028396</a>

Reviewed by Jer Noble.

To support entity-based rendering in LinearMediaPlayer, we must pass a video receiver endpoint from
LinearMediaPlayer to MediaPlayerPrivate so that it can be added as a target to AVPlayers and
AVSampleBufferVideoRenderers in the GPU process. This PR adds that plumbing (but saves actually
attaching the endpoint to AVFoundation objects for a later PR).

Note that a video receiver endpoint object contains an xpc_endpoint_t, which is not a type we can serialize in CoreIPC.
Like we do for other xpc_endpoint_ts, this PR sends it directly over the XPC connection from the UI
process to the GPU process.

* Source/WTF/wtf/Forward.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::setVideoReceiverEndpoint):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setVideoReceiverEndpoint):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/ios/VideoReceiverEndpoint.h: Renamed from Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.mm.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/GPUProcess/EntryPoint/Cocoa/XPCService/GPUServiceEntryPoint.mm:
(WebKit::initializeAuxiliaryProcess&lt;GPUProcess&gt;): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::GPUProcess):
(WebKit::GPUProcess::singleton):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(-[WKLinearMediaPlayerDelegate linearMediaPlayer:setVideoReceiverEndpoint:]):
* Source/WebKit/Platform/ios/VideoReceiverEndpointMessage.h: Added.
(WebKit::VideoReceiverEndpointMessage::messageName):
(WebKit::VideoReceiverEndpointMessage::processIdentifier const):
(WebKit::VideoReceiverEndpointMessage::playerIdentifier const):
(WebKit::VideoReceiverEndpointMessage::endpoint const):
* Source/WebKit/Platform/ios/VideoReceiverEndpointMessage.mm: Added.
(WebKit::VideoReceiverEndpointMessage::VideoReceiverEndpointMessage):
(WebKit::VideoReceiverEndpointMessage::decode):
(WebKit::VideoReceiverEndpointMessage::encode const):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.h: Renamed from Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.h.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.mm: Renamed from Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.mm.
(WebKit::handleLaunchServiceDatabaseMessage):
(WebKit::handleVideoReceiverEndpointMessage):
(WebKit::handleXPCEndpointMessage):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::setVideoReceiverEndpoint):
(WebKit::PlaybackSessionManagerProxy::setVideoReceiverEndpoint):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(WKSLinearMediaPlayer.setVideoReceiverEndpoint(_:)):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:

Canonical link: <a href="https://commits.webkit.org/274801@main">https://commits.webkit.org/274801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dd3089163128ca0095a0b39192e30aa152e44917

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40018 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19029 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42563 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35920 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42325 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16359 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33315 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16028 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34577 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13866 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13893 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43841 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33471 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36316 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35866 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39630 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39644 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14833 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12183 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37898 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16452 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46652 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8992 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16501 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9596 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16096 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
```diff
@@ -28,6 +28,7 @@
 #if PLATFORM(IOS_FAMILY) || (PLATFORM(MAC) && ENABLE(VIDEO_PRESENTATION_MODE))
 
 #include "PlatformMediaSession.h"
+#include "VideoReceiverEndpoint.h"
```
This should be forward-declarable.

aestes replied:
This header just contains a `using` declaration.

jernoble commented on Source/WebCore/platform/graphics/MediaPlayer.h:
```diff
@@ -41,6 +41,7 @@
 #include "SecurityOriginData.h"
 #include "Timer.h"
 #include "VideoPlaybackQualityMetrics.h"
+#include "VideoReceiverEndpoint.h"
```
Ditto.

aestes committed:
[Cocoa] Plumb a video receiver endpoint from LinearMediaPlayer to MediaPlayerPrivate
https://bugs.webkit.org/show_bug.cgi?id=269490
rdar://123028396

Reviewed by Jer Noble.

To support entity-based rendering in LinearMediaPlayer, we must pass a video receiver endpoint from
LinearMediaPlayer to MediaPlayerPrivate so that it can be added as a target to AVPlayers and
AVSampleBufferVideoRenderers in the GPU process. This PR adds that plumbing (but saves actually
attaching the endpoint to AVFoundation objects for a later PR).

Note that a video receiver endpoint object contains an xpc_endpoint_t, which is not a type we can serialize in CoreIPC.
Like we do for other xpc_endpoint_ts, this PR sends it directly over the XPC connection from the UI
process to the GPU process.

* Source/WTF/wtf/Forward.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::setVideoReceiverEndpoint):
* Source/WebCore/platform/graphics/MediaPlayer.h:
* Source/WebCore/platform/graphics/MediaPlayerPrivate.h:
(WebCore::MediaPlayerPrivateInterface::setVideoReceiverEndpoint):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.h:
* Source/WebCore/platform/ios/VideoReceiverEndpoint.h: Renamed from Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.mm.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/GPUProcess/EntryPoint/Cocoa/XPCService/GPUServiceEntryPoint.mm:
(WebKit::initializeAuxiliaryProcess<GPUProcess>): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::GPUProcess):
(WebKit::GPUProcess::singleton):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(-[WKLinearMediaPlayerDelegate linearMediaPlayer:setVideoReceiverEndpoint:]):
* Source/WebKit/Platform/ios/VideoReceiverEndpointMessage.h: Added.
(WebKit::VideoReceiverEndpointMessage::messageName):
(WebKit::VideoReceiverEndpointMessage::processIdentifier const):
(WebKit::VideoReceiverEndpointMessage::playerIdentifier const):
(WebKit::VideoReceiverEndpointMessage::endpoint const):
* Source/WebKit/Platform/ios/VideoReceiverEndpointMessage.mm: Added.
(WebKit::VideoReceiverEndpointMessage::VideoReceiverEndpointMessage):
(WebKit::VideoReceiverEndpointMessage::decode):
(WebKit::VideoReceiverEndpointMessage::encode const):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.h: Renamed from Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.h.
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCEndpointMessages.mm: Renamed from Source/WebKit/WebProcess/cocoa/HandleXPCEndpointMessages.mm.
(WebKit::handleLaunchServiceDatabaseMessage):
(WebKit::handleVideoReceiverEndpointMessage):
(WebKit::handleXPCEndpointMessage):
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceMain.mm:
(WebKit::XPCServiceEventHandler):
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::setVideoReceiverEndpoint):
(WebKit::PlaybackSessionManagerProxy::setVideoReceiverEndpoint):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(WKSLinearMediaPlayer.setVideoReceiverEndpoint(_:)):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:

Canonical link: https://commits.webkit.org/274801@main
